### PR TITLE
[JENKINS-30435][JENKINS-30436] Updating proxies on template project rename and non-existing project name validation

### PR DESCRIPTION
**Pull request by oleksandr-horobets**

### CODE REVIEW

File: src/main/java/hudson/plugins/templateproject/ItemListenerImpl.java [hunk 1]
```diff
@@ -47,11 +62,46 @@ public void onLoaded() {
...
 	}
 
+	@Override
+	public void onRenamed(Item item, String oldName, String newName) {
+		for (AbstractProject project : hudson.getAllItems(AbstractProject.class)) {
+			updateHandler.updateScm(project, oldName, newName);
+
+			boolean updatedItems = false;
+
+			if(project instanceof Project){
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/ItemListenerImpl.java]: Bad approach. There may be other implementations of abstract project, which are being ignored here

File: src/main/java/hudson/plugins/templateproject/ItemListenerImpl.java [hunk 2]
```diff
@@ -47,11 +62,46 @@ public void onLoaded() {
 		}
 	}
 
+	@Override
+	public void onRenamed(Item item, String oldName, String newName) {
+		for (AbstractProject project : hudson.getAllItems(AbstractProject.class)) {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/ItemListenerImpl.java]: This may cause a serious performance degradation on renames. But there is no other obvious way since we don't store backward links from templated projects.
I also doubt such rename hook will work correctly with folders. CC @jglick

File: src/main/java/hudson/plugins/templateproject/ProxyBuilder.java [hunk 3]
```diff
@@ -4,21 +4,16 @@
...
-import hudson.model.AbstractBuild;
-import hudson.model.AbstractProject;
-import hudson.model.Action;
-import hudson.model.BuildListener;
-import hudson.model.DependecyDeclarer;
-import hudson.model.DependencyGraph;
-import hudson.model.Hudson;
-import hudson.model.Item;
-import hudson.model.Project;
+import hudson.model.*;
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/ProxyBuilder.java]: Bad practice for the non-test code

File: src/main/java/hudson/plugins/templateproject/ProxyBuilder.java [hunk 4]
```diff
@@ -45,16 +34,22 @@ public String getProjectName() {
 		return projectName;
 	}
 
-	public String getExpandedProjectName(AbstractBuild<?, ?> build) {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/ProxyBuilder.java]: Why do you delete it? Causes binary compatibility issues from what I see

File: src/main/java/hudson/plugins/templateproject/ProxyBuilder.java [hunk 5]
```diff
@@ -45,16 +34,22 @@ public String getProjectName() {
...
-		return TemplateUtils.getExpandedProjectName(projectName, build);
+	public List<Builder> getProjectBuilders(AbstractBuild<?, ?> build) {
+		return getBuildersFrom(TemplateUtils.getInstance().getProject(getProjectName(), build));
 	}
 
-	public List<Builder> getProjectBuilders(AbstractBuild<?, ?> build) {
-		AbstractProject p = TemplateUtils.getProject(getProjectName(), build);
+	public List<Builder> getProjectBuilders() {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/ProxyBuilder.java]: same as above

File: src/main/java/hudson/plugins/templateproject/ProxyBuilder.java [hunk 6]
```diff
@@ -104,18 +99,25 @@ public FormValidation doCheckProjectName(@AncestorInPath AccessControlled anc, @
...
+					return false;
+				}
+				listener.getLogger().println("[TemplateProject] Successfully performed builders from: '" + project.getFullDisplayName() + "'");
 			}
-			listener.getLogger().println("[TemplateProject] Successfully performed builders from: '" + p.getFullDisplayName() + "'");
+
+			return true;
+
+		} else {
+			listener.getLogger().printf("[TemplateProject] FAILED builders loading, couldn't find project: '%s' %n", getProjectName());
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/ProxyBuilder.java]: It may break some use-cases, but I also vote for having it as a default behavior

File: src/main/java/hudson/plugins/templateproject/ProxyPublisher.java [hunk 7]
```diff
@@ -43,14 +43,6 @@ public String getProjectName() {
 		return projectName;
 	}
 
-	public String getExpandedProjectName(AbstractBuild<?, ?> build) {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/ProxyPublisher.java]: Binary compatibility issue?

File: src/main/java/hudson/plugins/templateproject/ProxyPublisher.java [hunk 8]
```diff
@@ -60,43 +52,54 @@ public boolean needsToRunAfterFinalized() {
...
+
+		if(project != null) {
+			for (Publisher publisher : project.getPublishersList()) {
+				if (!publisher.prebuild(build, listener)) {
+					return false;
+				}
 			}
+			return true;
+		} else {
+			return false;
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/ProxyPublisher.java]: Makes sense to print an error to the build listener

File: src/main/java/hudson/plugins/templateproject/ProxySCM.java [hunk 9]
```diff
@@ -52,19 +47,12 @@ public String getProjectName() {
...
-
-	// Primarily used for polling, not building.
-	public AbstractProject<?, ?> getProject() {
-		return TemplateUtils.getProject(projectName, null);
-	}
-
 	public SCM getProjectScm(AbstractBuild<?, ?> build) {
-		try {
-			return TemplateUtils.getProject(projectName, build).getScm();
-		} catch (Exception e) {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/ProxySCM.java]: Exception may mean the runtime exception like NPE? Are you sure you want to delete this catch?

File: src/main/java/hudson/plugins/templateproject/TemplateUtils.java [hunk 10]
```diff
@@ -1,38 +1,52 @@
...
-		if (build != null) {
-			pName = TemplateUtils.getExpandedProjectName(projectName, build);
+	private final Hudson hudson;
 
-			if (Hudson.getInstance().getItemByFullName(pName) == null) {
-				logger.info("[TemplateProject] Template Project '" + pName + "' not found. Skipping.");
-			}
-		}
+	private TemplateUtils(){
+		this(Hudson.getInstance());
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/TemplateUtils.java]: This approach is wrong. Your class may be initialized during Jenkins startup. iIn such case Jenkins instance will be null

File: src/main/java/hudson/plugins/templateproject/TemplateUtils.java [hunk 11]
```diff
@@ -1,38 +1,52 @@
...
 public class TemplateUtils {
-	private static final Logger logger = Logger.getLogger("TemplateProject");
 
-	public static AbstractProject<?, ?> getProject(String projectName, AbstractBuild<?, ?> build) {
-		String pName = projectName;
+	private static TemplateUtils instance;
 
-		if (build != null) {
-			pName = TemplateUtils.getExpandedProjectName(projectName, build);
+	private final Hudson hudson;
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/TemplateUtils.java]: Use jenkins.model.Jenkins

File: src/main/java/hudson/plugins/templateproject/TemplateUtils.java (hunk 10 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/TemplateUtils.java]: I would recommend not to cache the value. Get Instance is fast in the common implementation for the runtime

File: src/main/java/hudson/plugins/templateproject/UpdateHandler.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+            ProxySCM proxySCM = (ProxySCM) scm;
+
+            if (oldName.equals(proxySCM.getProjectName())) {
+                try {
+                    project.setScm(new ProxySCM(newName));
+                } catch (IOException e) {
+                    LOGGER.log(Level.SEVERE, "[TemplateProject] Unable to update SCM template reference", e);
+                }
+            }
+        }
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/UpdateHandler.java]: else?

File: src/main/java/hudson/plugins/templateproject/TemplateUtils.java (hunk 10 above)
Comment (oleksandr-horobets) [inline src/main/java/hudson/plugins/templateproject/TemplateUtils.java]: it was done to simplify testing of this class, since I need to mock Hudson instance for several tests. Is there some safe way to obtain Jenkins instance only when it's available? Something like blocking method?

File: src/main/java/hudson/plugins/templateproject/TemplateUtils.java (hunk 11 above)
Comment (oleksandr-horobets) [inline src/main/java/hudson/plugins/templateproject/TemplateUtils.java]: There are a lot of Hudson usages in this plugin. Does it make sense to change it to Jenkins in all places? Won't it break compatibility for Hudson users?

File: src/main/java/hudson/plugins/templateproject/TemplateUtils.java (hunk 10 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/TemplateUtils.java]: Just a good practice for the new code. Refactoring may make sense, but it's very optional

File: src/main/java/hudson/plugins/templateproject/TemplateUtils.java (hunk 11 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/TemplateUtils.java]: Just a good practice for the new code. Refactoring may make sense, but it's very optional

File: src/main/java/hudson/plugins/templateproject/ItemListenerImpl.java (hunk 2 above)
Comment (oleksandr-horobets) [inline src/main/java/hudson/plugins/templateproject/ItemListenerImpl.java]: yes, for sure it will be slow. Let me try to implement some kind of caching.

File: src/main/java/hudson/plugins/templateproject/TemplateUtils.java (hunk 10 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/TemplateUtils.java]: No, there is no blocking methods for such purpose. + there are test frameworks starting multiple Jenkins instances within one JVM (they rely on TLS-based holder in Jenkins.getInstance())

File: src/main/java/hudson/plugins/templateproject/ItemListenerImpl.java (hunk 2 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/ItemListenerImpl.java]: I would leave it as is now and create a reference management engine later. There are project actions, which allow to persist and manage the data. This solution will be pretty complex, but useful

File: src/main/java/hudson/plugins/templateproject/ProxyBuilder.java (hunk 4 above)
Comment (oleksandr-horobets) [inline src/main/java/hudson/plugins/templateproject/ProxyBuilder.java]: I have a suggestion that this method was declared as public accidentally. I don't see any particular reason to expose it as public. There are no usages in this project, but I'm not sure about other plugins. Is there a way to find what plugins depends on this plugin?

File: src/main/java/hudson/plugins/templateproject/TemplateUtils.java (hunk 10 above)
Comment (oleksandr-horobets) [inline src/main/java/hudson/plugins/templateproject/TemplateUtils.java]: Oleg, do you mean http://javadoc.jenkins-ci.org/?org/jvnet/hudson/test/JenkinsRule.html?

File: src/main/java/hudson/plugins/templateproject/ProxySCM.java (hunk 9 above)
Comment (oleksandr-horobets) [inline src/main/java/hudson/plugins/templateproject/ProxySCM.java]: not sure where NPE could occur. If project is null NullSCM will be returned

File: src/main/java/hudson/plugins/templateproject/UpdateHandler.java (hunk 12 above)
Comment (oleksandr-horobets) [inline src/main/java/hudson/plugins/templateproject/UpdateHandler.java]: Oleg, do you mean multiple SCM support?

File: src/main/java/hudson/plugins/templateproject/ItemListenerImpl.java (hunk 1 above)
Comment (oleksandr-horobets) [inline src/main/java/hudson/plugins/templateproject/ItemListenerImpl.java]: I don't like this approach too, but how can I obtain project builders using only AbstractProject?

File: src/main/java/hudson/plugins/templateproject/UpdateHandler.java (hunk 12 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/UpdateHandler.java]: I've noticed that ProxySCM is a member of this plugin. BTW you're right that MultipleSCM is a potential case.

File: src/main/java/hudson/plugins/templateproject/TemplateUtils.java (hunk 10 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/TemplateUtils.java]: No, the test framework I reference is no closed-source unfortunately. BTW it was just a corner-case example.

In the case of your tests I would say that you could use <code>JenkinsRule</code> instead of mocking. Such functional tests take more time, but they provide a better coverage.

File: src/main/java/hudson/plugins/templateproject/ProxySCM.java (hunk 9 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/ProxySCM.java]: NPE is just an example. IllegalStateException, IllegalArgumentException and other such guys are also theoretically possible. I have no idea what this <code>catch</code> clause expects to get, but I would avoid deleting it without prior investigation.

File: src/main/java/hudson/plugins/templateproject/ProxyBuilder.java (hunk 4 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/ProxyBuilder.java]: It is possible, but unfortunately it does not help much. The API method may be used in closed-source plugins and in System Groovy Scripts in Jenkins, so a plugin check is not enough to guarantee the compatibility.

If you feel the method should not be public, do the following:
* Mark the method as <code>@Deprecated</code>
* Add the <code>@Restricted(NoExternalUse.class)</code> annotation. It will prevent dependent plugins from compiling with the new plugin versions if they use this method

File: src/main/java/hudson/plugins/templateproject/ItemListenerImpl.java (hunk 1 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/templateproject/ItemListenerImpl.java]: Agreed. I would recommend to add a logging on the INFO level in such case. This approach may become obsolete when Template projects start referencing their users somehow.

File: src/main/java/hudson/plugins/templateproject/ProxyBuilder.java (hunk 4 above)
Comment (Brantone) [inline src/main/java/hudson/plugins/templateproject/ProxyBuilder.java]: +1 for annotations.

**Comment by Brantone** [comment]
@oleksandr-horobets  : Will some of these recommendations be implemented?

**Comment by Brantone** [comment]
Still un-implemented and now with conflicts to resolve :(
The tests are nice, but lots of recommendations that should be corrected

**Comment by oleksandr-horobets** [comment]
@Brantone I'm working on comments now, hope will be done soon:)
